Add a bit about various type system concepts

mark-i-m opened the pull request:
A lot of this is taken straight from Niko's writings here:
- https://gist.github.com/nikomatsakis/fe440aec437f7e8ad4cce8319954a973
- https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/reintegrating.20chalk.20into.20rustc/near/190518417

r? @nikomatsakis 

resolves #177 I think

spastorino commented:
This looks good!

tshepang commented on src/appendix/background.md:
````diff
@@ -79,7 +79,69 @@ cycle.
+ +---------+
+```
+
+We can do the dataflow analysis as follows: we will start of with a flag `init`
````
```suggestion
We can do the dataflow analysis as follows: we will start off with a flag `init`
```

tshepang commented on src/appendix/background.md:
```diff
@@ -79,7 +79,69 @@ cycle.
+initialized. So at the end of (B), `init = true`.
+
+Block (C) is where things get interesting. Notice that there are two incoming
+edges from (A) and (B), corresponding to whether `some_cond` is true or not.
```
```suggestion
edges, one from (A) and one from (B), corresponding to whether `some_cond` is true or not.
```
that looks more clear

tshepang commented on src/appendix/background.md:
```diff
@@ -79,7 +79,69 @@ cycle.
+(C) if it is true in both (A) and (B), i.e., both parents of (C). Since, `init
+= false` in (A), `init = false` in (C), and we can report to the user that "`x`
+may not be initialized before use".
 
```
I don't understand this last sentence... can you rephrase.

mark-i-m replied:
I rephrased it and expanded a bit. Let me know what you think.

tshepang commented on src/appendix/background.md:
```diff
@@ -184,6 +184,30 @@ bound, it is well-typed: `∀ F, 'a: (F: Fn(&'a u8)) => well_typed(foo)`.
 
+## What is a DeBruijn Index?
+
+DeBruijn indices are a way of representing the which variables are bound in
+which binders using only integers. They were [originally invented][wikideb] for
```
```suggestion
DeBruijn indices are a way of representing which variables are bound in
which binders using only integers. They were [originally invented][wikideb] for
```

tshepang commented on src/early-late-bound.md:
```diff
@@ -0,0 +1,122 @@
+
+Early-bound parameters in Rustc are identified by an index, stored in the
+[`ParamTy`] struct for types or the [`EarlyBoundRegion`] struct for lifetimes.
+The index counts from the outermost declaration in scope. This meansthat as you
```
```suggestion
The index counts from the outermost declaration in scope. This means that as you
```

tshepang commented on src/early-late-bound.md:
```diff
@@ -0,0 +1,122 @@
+
+Moreover, a reference to a late-bound lifetime is written `^0.a`:
+
+    - The `0` is the index; it identifies that this lifetime is bound in the
```
this renders as a code block... is it intended

nikomatsakis commented on src/appendix/background.md:
````diff
@@ -79,16 +79,150 @@ cycle.
+Another example:
+
+```rust,ignore
+fn foo<'a>(_: &'a usize)
````
This is also universal quantification -- for any lifetime `'a` that the caller chooses, if you supply a `&'a usize`, `foo` will succeed.

nikomatsakis commented on src/appendix/background.md:
````diff
@@ -79,16 +79,150 @@ cycle.
+fn foo<'a>(_: &'a usize)
+```
+This function claims that there is some lifetime `'a` (determined by the
+caller) such that it is well-typed: `∃ 'a: well_typed(foo)`.
````
Existential would be like a `dyn Debug` type, which is kind of `exists<T> { T: Debug }`. i.e., you have a value of some type `T` that is `Debug` (but there is some concrete type `T`).

nikomatsakis commented on src/appendix/background.md:
````diff
@@ -79,16 +79,150 @@ cycle.
+where for<'a> F: Fn(&'a u8),
+```
+This function claims that for all lifetimes `'a` and types `F` satisfying the
+bound, it is well-typed: `∀ F, 'a: (F: Fn(&'a u8)) => well_typed(foo)`.
````
Not quite. This function claims that it is well-typed for al types `F` where:

* for all lifetimes `'a`, `F` implements `Fn(&'a u8)`

nikomatsakis replied:
```
forall<F> {
    if (forall<'a> { F: Fn(&'a u8) }) {
        /* foo::<F> is well typed */
    }
}
```

nikomatsakis commented on src/appendix/background.md:
````diff
@@ -79,16 +79,150 @@ cycle.
+                // index of `y` is 1 because it is bound 1 level up
+    }
+}
+```
````
yep

nikomatsakis commented on src/early-late-bound.md:
```diff
@@ -0,0 +1,122 @@
+# Early and Late Bound Variables
+
+A generic type may be [_universally_ or _existentially_ quantified][quant]. For
```
I'm not sure this is true. All the generics in Rust correspond to universal quantification.

nikomatsakis commented on src/early-late-bound.md:
````diff
@@ -0,0 +1,122 @@
+fn foo<T>()
+```
+This function claims that the function is well-typed for all types `T`. To use
+the chalk notation: `forall<T> foo<T>`.
````
This seems to duplicate the material above.

nikomatsakis replied:
I'm also not sure how relevant it is here

nikomatsakis commented on src/early-late-bound.md:
```diff
@@ -0,0 +1,122 @@
+We cannot treat `'a`, `'b`, and `T` in the same way.  Types in Rust can't have
+`for<T> { .. }`, only `for<'a> {...}`, so whenever you reference `foo` the type
+you get back can't be `for<'a, 'b, T> fn(&'a u32, y: &'b T)`. Instead, the `T`
+must be substituted early. In particular, you have:
```
This seems correct, yes.

nikomatsakis commented on src/early-late-bound.md:
```diff
@@ -0,0 +1,122 @@
+type to find out what is bound at the given level and so forth. The advantage
+is primarily that, when constructing types from Rust syntax, if we encounter
+anonymous regions like in `fn(&u32)`, we just create a fresh index and don't have
+to update the binder.
```
correct

nikomatsakis commented on src/appendix/background.md:
```diff
@@ -79,16 +79,150 @@ cycle.
+We only discussed a forwards analysis, but backwards dataflow analysis is also
+useful. For example, rather than starting from block (A) and moving backwards,
+we might have started with the usage of `x` and moved backwards to try to find
+its initialization.
```
seems good!

tshepang commented on src/early-late-bound.md:
```diff
@@ -0,0 +1,122 @@
+
+## Early-bound parameters
+
+Early-bound parameters in Rustc are identified by an index, stored in the
```
```suggestion
Early-bound parameters in rustc are identified by an index, stored in the
```

tshepang commented on src/early-late-bound.md:
```diff
@@ -0,0 +1,122 @@
+## Late-bound parameters
+
+Late-bound parameters in `rustc` are handled quite differently (they are also
+specialized to lifetimes, since right now only late-bound lifetimes are
```
```suggestion
specialized to lifetimes since, right now, only late-bound lifetimes are
```

tshepang commented on src/appendix/background.md:
```diff
@@ -79,16 +79,150 @@ cycle.
+[rice]: https://en.wikipedia.org/wiki/Rice%27s_theorem
+
+Generally, in dataflow analyses, if a block has multiple parents (like (C) in
+our example), its dataflow value will be some function of all it's parents (and
```
```suggestion
our example), its dataflow value will be some function of all its parents (and
```

mark-i-m commented:
@nikomatsakis I think I addressed your comments. Could you take a look at the beginning of the early/late bounds chapter again? (basically the last commit)...

nikomatsakis commented on src/early-late-bound.md:
````diff
@@ -1,46 +1,13 @@
-```
-
+In Rust, generic types are generally [_universally_ quantified][quant].
 Notice, however, that in Rust, we don't have (at the language level)
````
Oops, sorry @mark-i-m for the delay -- but this doesn't seem right to me. The type `for<'a> fn(&'a u32)`, for example, is a universally quantified type, as is `dyn for<'a> Foo<&'a u32>` or something like that. It is true that the quantifiers are attached to specific kinds of types (`fn` types, `dyn` types), and we don't have some general purpose "type quantification" mechanism. 

(As far as early- vs late- splits go, any region that is bound in a type is late-bound.)

mark-i-m replied:
Hi @nikomatsakis, I adjusted this paragraph again in the latest commit, but I'm not sure if I'm just misunderstanding you...

nikomatsakis commented on src/early-late-bound.md:
```diff
@@ -0,0 +1,90 @@
+# Early and Late Bound Variables
+
+In Rust, generic types are generally [_universally_ quantified][quant].
```
well, I think I would phrase this like so

In Rust, item definitions (like `fn`) can often have generic parameters, which are always [_universally_ quantified][quant]. That is, if you have a function like

```rust
fn foo<T>(x: T) { }
``` 

this function is defined "for all T" (not "for some specific T", which would be existentially quantified).

While Rust *items* can be quantified over types, lifetimes, and constants, the types of values in Rust are only ever quantified over lifetimes. So you can have a type like `for<'a> fn(&'a u32)`, which represents a function pointer that takes a reference with any lifetime, or `for<'a> dyn Trait<'a>`, which is a `dyn` trait for a trait implemented for any lifetime; but we have no type like `for<T> fn(T)`, which would be a function that takes a value of *any type* as a parameter. This is a consequence of monomorphization -- to support a value of type `for<T> fn(T)`, we would need a single function pointer that can be used for a parameter of any type, but in Rust we generate customized code for each parameter type. 

One consequence of this asymmetry is a weird split in how we represesent some generic types: _early-_ and _late-_ bound parameters.

mark-i-m replied:
I replaced the intro with this (let me know if I didn't replace enough). Sorry, I'm only slowly starting to understand this distinction; you're intro helps a lot.

mark-i-m commented:
Oh, huh? I thought I had already merged it! Sorry!
